cmd/snap-bootstrap/hybrid_users_import.go

```diff
@@ -0,0 +1,582 @@
+		// import the primary groups that those users are in. these will most likely
+		// all be unique, but there is a chance that users share a primary group.
+		for _, u := range importUsers {
+			if u.gid == g.gid {
+				return true
+			}
+		}
+
+		return false
+	}
+}
+
+// mergeAndWriteLoginFiles considers the etc/passwd, etc/shadow, etc/group, and
+// etc/gshadow files from the given importRoot and baseRoot directories, merges
+// them together, and writes the merged files into the given outputDir.
+//
+// Only non-system (uid and gid < 1000) users that are members of any of the
+// groups in the targetGroups slice are imported from the importRoot directory.
+// All users and groups from the baseRoot are imported. As a special case, the
+// root user is always imported from importRoot.
+//
+// If there are any conflicts between the users and groups in the importRoot and
+// baseRoot directories, the importRoot users and groups are used.
+func mergeAndWriteLoginFiles(importRoot, baseRoot string, targetGroups []string, outputDir string) error {
+	baseUsers, err := parseUsers(baseRoot, nil)
+	if err != nil {
+		return err
+	}
+
+	baseGroups, err := parseGroups(baseRoot, nil)
+	if err != nil {
+		return err
+	}
+
+	loginShells, err := parseShells(importRoot)
+	if err != nil {
+		return err
+	}
+
+	importUsers, err := parseUsers(importRoot, userFilter(
+		targetGroups,
+		baseUsers,
+		baseGroups,
+		loginShells,
+	))
+	if err != nil {
+		return err
+	}
+
+	importGroups, err := parseGroups(importRoot, groupFilter(importUsers))
+	if err != nil {
+		return err
+	}
+
+	// update the passwd and shadow files to contain the lines for the users
+	// that we're importing
+	err = mergeAndWriteUserFiles(baseUsers, importUsers, outputDir)
+	if err != nil {
+		return err
+	}
+
+	// update the group and gshadow files to add our users to any existing
+	// groups, and add any new primary groups that we're importing.
+	if err := mergeAndWriteGroupFiles(baseGroups, importUsers, importGroups, outputDir); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// mergeAndWriteUserFiles merges the base users with the given users to be
+// imported. The caller must ensure that the UIDs and GIDs of any of the given
+// users do not conflict with any of the given base users. The merged files are
+// written into the given output directory.
+func mergeAndWriteUserFiles(baseUsers map[string]user, importUsers map[string]user, outputDir string) error {
+	var passwdBuffer bytes.Buffer
+	var shadowBuffer bytes.Buffer
+	for name, user := range baseUsers {
+		if name != user.name {
+			return fmt.Errorf("internal error: user entry inconsistent with parsed data")
+		}
+
+		// if there is a conflict in the existing file, we will use the imported
+		// entry instead of the base one. this is consistent with how we handle
+		// groups, so any conflicting users will always be fully replaced.
+		if _, ok := importUsers[name]; ok {
```

Comment:
if I understand correctly we skip importUsers if there's a conflict in the uid but we prefer them if the conflict is in the name? shouldn't we just consistently skip them in either case?

Reply:
Changed. Note that this does require some special handling of the root user, since we do want to import that one always.